httpcore5/src/main/java/org/apache/hc/core5/http/impl/io/DefaultResponseOutOfOrderStrategy.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.core5.http.impl.io;
+
+import org.apache.hc.core5.http.ClassicHttpRequest;
+import org.apache.hc.core5.http.io.ResponseOutOfOrderStrategy;
+
+/**
+ * Default {@link ResponseOutOfOrderStrategy} implementation which checks for premature responses every 8kb based
+ * on testing using values between 4 KiB and 128 KiB. This is optimized for correctness, and results in a maximum
+ * upload speed of 8 MiB/s.
```

Comment:
Didn't you say you lose 8 MiB/s on a 100 MBit/s link?

Reply:
No, unfortunately it limits any link to at most 8 MiB/s. In the good case, the server does not respond until clients have successfully transferred all request bytes. The check must spend 1ms every 8KiB transferred because the full timeout is reached when no bytes are available. Assuming instantaneous transfer speed, the out of order check reduces us to 8KiB per millisecond which is roughly equivalent to 8MiB/second.

Reply:
There's an interesting case where all writes use buffers larger than 8k, which results in a 1ms sleep for each write, rather than 8k chunks. I don't think this can occur using chunked transfer-encoding, but may be problematic for content-length request entities. I will file a separate issue to discuss segmenting writes based on the response check interval.

Reply:
I have some request log aggregation graphs from an internal production service which shows 40-50mb uploads dropping from 6 seconds p50 to 1 second p50 after I disabled the early response check by subclassing DefaultBHttpClientConnection. I'll try to clean up the graphs so I can post them, I found it interesting that they very closely match the 8 MiB/second napkin math.

Reply:
You are of course right, this early response chunk size should be a multiple of the buffer size used to write the entity. We cannot access this buffer size to size dynamically.
I also ask myself whether 2 KiB for chunked TE isn't too small, but this is another discussion.